### ISSUE TITLE
Additional <<foo slot syntax

### ISSUE THREAD

**Issue by mrjbq7** [factor/factor#2839]
With this patch:

```diff
diff --git a/core/slots/slots.factor b/core/slots/slots.factor
index 0618975b9b..ea06b8fde4 100644
--- a/core/slots/slots.factor
+++ b/core/slots/slots.factor
@@ -65,6 +65,15 @@ M: object reader-quot
     "<<" append "accessors" create-word
     dup t "writer" set-word-prop ;
 
+: getter-word ( name -- word )
+    "<<" prepend "accessors" create-word ;
+
+: define-getter ( name -- )
+    dup getter-word dup deferred? [
+        swap reader-word 1quotation [ keep ] curry
+        ( object -- value object ) define-inline
+    ] [ 2drop ] if ;
+
 ERROR: bad-slot-value value class ;
 
 GENERIC: instance-check-quot ( obj -- quot )
@@ -126,6 +135,7 @@ M: object writer-quot
 
 : define-slot-methods ( class slot-spec -- )
     [ define-reader ]
+    [ name>> define-getter drop ]
     [
         dup read-only>> [ 2drop ] [
             [ name>> define-setter drop ]
@@ -133,7 +143,7 @@ M: object writer-quot
             [ define-writer ]
             2tri
         ] if
-    ] 2bi ;
+    ] 2tri ;
 
 : define-accessors ( class specs -- )
     [ define-slot-methods ] with each ;

```

Then you can do some stuff:

```
IN: scratchpad TUPLE: foo a b c ;

IN: scratchpad 1 2 3 foo boa

--- Data stack:
T{ foo f 1 2 3 }

IN: scratchpad <<a dup c>>


--- Data stack:
1
T{ foo f 1 2 3 }
3
```

Some comments on Discord about this being:

* good (``<<a dup c>>`` better than ``[ a>> ] [ ] [ c>> ] tri``)
* medium (why not just ``get[ a c ]``)
* bad (why do we use ``<<`` and ``>>`` anyway, how about slot objects and something like forth's ``@`` and ``!`` get/set syntax)

**Comment by mrjbq7** [comment]
Thanks @tgunr for the idea!

**Comment by tgunr** [comment]
I like the 'why do we use << >> anyway?'  seems to me  `slot name @' 'value slot !' is more clear. When I first discovered accessors I was confused as to these auto created words and why <<slot was missing.

**Comment by Capital-EX** [comment]
So would it be something like the following syntax:
```factor
TUPLE: foobar
   a b c ;
   

1 2 3 foobar boa [ 10 a ! ] [ a foobar @ . ] bi
```

If so, how would we keep slot names from polluting the namespace? Would they be able to be qualified via their tuple type (`foobar:a`)?

**Comment by erg** [comment]
my default image size (that loads crap in .factor-boot-rc) went from to 119,531,384 to 120,288,680 == 757,296 for the accessor. hm

**Comment by tgunr** [comment]
Probably too much code out there to implement something new for slots. However, something like `foobar:a @ ` and `99 foobar:a !` seems attractive enough and would prevent the namespace pollution @Capital-EX mentioned. But the use of "!" as comment rears its ugly head and was an unfortunate choice IMHO. Fetch and store are so common they deserve a single character and in my opinion the use of extended character set should be used, eg: ¡ = store, § = fetch, or » = store and « = fetch. The use of the option key soon becomes common once you get used to the idea.

**Comment by erg** [comment]
What if we had two generic words, one to get slots and one to set them, and the dual getter and setter were not full on words? Would this cut the code size down?

**Comment by mrjbq7** [comment]
I think we wouldn't do ``foobar:a`` because that prevents "duck typing" like our current ``a>>`` accessor.

I was thinking more like:

```
SLOT: a ! declared somewhere implicit like our accessors vocab

...

a @
```

I agree that the comment character prevents using ``!`` without hassle, and before we make any changes, we should probably think about getting/setting of slots, locals, namespace variables, thread-local variables, global variables, memcached values, etc.

Maybe there's a unifying concept or unifying language-consistency for all of these.

**Comment by mrjbq7** [comment]
> What if we had two generic words, one to get slots and one to set them, and the dual getter and setter were not full on words? Would this cut the code size down?

This is what we have now:

```factor
GENERIC: name>> ( object -- value )
GENERIC: name<< ( value object -- )
: >>name ( object value -- object ) over name<< ; inline
```

Proposed addition:

```factor
: <<name ( object -- value object ) [ >>name ] keep ; inline
```

**Comment by tgunr** [comment]
You mean you would have to explicitly define the getter and setter if you planned to use them?

**Comment by tgunr** [comment]
I agree, those forms are so common there should be a way to generalize them in some kind of syntax form.
